### PR TITLE
Adds --vram_O option

### PR DESCRIPTION
**Pull request by tobbelobb**
### PR is related to a problem

Yes, I wanted to run on 8 GB GPU (a 3070), but couldn't do it without some adjustments.

Enabling the provided memory saving sd config solves CUDA out of memory on my particular hardware setup.

The config's changes and their trade-offs:
  - Loads the model weights directly in half precision (16f instead of 32f). This might reduce output quality, not sure.
  - Enables attention slicing. This slows down inference times by about 10%.

### Alternatives I've considered

I looked up huggingface docs for memory savnig config alternatives.
https://huggingface.co/docs/diffusers/v0.14.0/en/optimization/fp16

To be honest I just implemented each optimization suggestion from the top and downwards until it worked.

After adding `self.unet.set_attention_slice("auto"), I got total GPU memory usage of 7132MiB, which was low enough for my GPU. That number was gotten from a test run like this:

```bash
$ python main.py --text "a hamburger" --workspace trial -O --memory_saving_sd_config
```

This works as expected and creates a hamburger in ~39 minutes.

### CODE REVIEW

**Comment by tobbelobb** [comment]
I realized that if I used more of the optimizations listed in the huggingface docs, I might get away with --w 128 and --h 128. I pushed the second optimization level as an option since it dramatically improved quality of output while still using below 8G of VRAM.

The left hamburger was generated in 39 min with
```bash
$ python main.py --text "a hamburger" --workspace trial -O --vram_O 1
```
the right one was generated in 65 min with
```bash
$ python main.py --text "a hamburger" --workspace trial -O --vram_O 2 --w 128 --h 128
```

Screenshot from 2023-03-22 00-36-47

Maybe this could allow those of you with more VRAM to run with `--w 512 --h 512`? I guess that would produce even nicer hamburgers.

File: sd.py [hunk 1]
```diff
@@ -54,16 +54,25 @@ def __init__(self, device, sd_version='2.1', hf_key=None):
...
+            pipe = StableDiffusionPipeline.from_pretrained(model_key, torch_dtype=torch.float16)
+            if vram_O > 1:
+                pipe.enable_sequential_cpu_offload()
+            pipe.enable_attention_slicing(1)
+            self.vae = pipe.vae
+            self.tokenizer = pipe.tokenizer
+            self.text_encoder = pipe.text_encoder
+            self.unet = pipe.unet
+            self.scheduler = DDIMScheduler.from_pretrained(model_key, subfolder="scheduler", torch_dtype=torch.float16)
+        else:
```
Comment (ashawkey) [inline sd.py:67]: we can remove the following lines and always use pipeline.

File: sd.py [hunk 2]
```diff
@@ -54,16 +54,25 @@ def __init__(self, device, sd_version='2.1', hf_key=None):
             raise ValueError(f'Stable-diffusion version {self.sd_version} not supported.')
 
         # Create model
-        self.vae = AutoencoderKL.from_pretrained(model_key, subfolder="vae").to(self.device)
-        self.tokenizer = CLIPTokenizer.from_pretrained(model_key, subfolder="tokenizer")
-        self.text_encoder = CLIPTextModel.from_pretrained(model_key, subfolder="text_encoder").to(self.device)
-        self.unet = UNet2DConditionModel.from_pretrained(model_key, subfolder="unet").to(self.device)
+        if vram_O > 0:
+            pipe = StableDiffusionPipeline.from_pretrained(model_key, torch_dtype=torch.float16)
```
Comment (ashawkey) [inline sd.py:58]: consider using opt.fp16 to control data type? like passing in opt.fp16 in __init__ and use `torch_dtype=torch.float16 if fp16 else torch.float32`

**Comment by ashawkey** [comment]
The using of pipeline is neat, I think we can make that the default way (making vram_O == 1 the default?).
Also, since there is already an `--fp16` option, maybe we can use it to control SD loading dtype.
Would you mind helping to make these changes? Or I could merge now and make a new commit.

It's interesting that the mesh quality can be improved with higher resolution!

**Comment by tobbelobb** [comment]
Yes it's quite interesting. I continued stepping through the huggingface optimization guide today, trying to push for w512 h512 renders on my hw. I got only to w300 h300 but it looked really great. It also took only 55 min.

Here are some burgers. Left to right: 64x64, 128x128, 256x256, 300x300
Screenshot from 2023-03-23 01-56-53

I also managed to use the Trace optimization mentioned in the docs page. It uses some more vram, but lets me generate a 200x200 (high-ish quality) burger in ~20 min, down from ~30 without the tracing.

It will take some time to disentangle and clean up. I like the idea to just use the `--fp16` option. I think `--fp16` feels like a default to new users since it's included in the `-O` option that is mentioned in the first usage example in the readme. Maybe we should rename `--fp16` to `--save_vram` or something similar, since it will do more than just reduce precision.

It makes sense to keep the non-vram-optimized mode around, since we won't get further quality gains by going beyond 512x512 anyways.

About the tracing feature and code. It requires a bit of effort to set up but it saves so much time that it will be my default way of using stable-dreamfusion. Should I include the somewhat messy trace code, and explain the process of using it in the readme, or should I just leave it out of this PR?

File: sd.py (hunk 2 above)
Comment (tobbelobb) [reply sd.py:58]: This looks nice. Will do it like this. Thanks

File: sd.py (hunk 1 above)
Comment (tobbelobb) [reply sd.py:67]: :ok_hand:

**Comment by tobbelobb** [comment]
I included everything into this PR so you can look at it before making up your mind.

**Comment by ashawkey** [comment]
I prefer to separate speed and memory optimization options, as `--fp16` can also boost speed, while `--vram_O` focuses on reducing memory even at the cost of speed (like cpu offloading). I'll merge the current ones and do some slight modifications.
Thanks again for your contribution!

**Comment by tobbelobb** [comment]
Thanks for creating and maintaining this awesome repo btw. It really blew my mind to realize this existed :pray:

**Pull request merged**